File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/Scripting.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface ScriptingLib: Library {
    fun Scripting_destroy(handle: Pointer)
    fun Scripting_run_lua_script(path: Slice): ResultPointerInt
    fun Scripting_run_js_script(path: Slice): ResultPointerInt
    fun Scripting_run_script(path: Slice): ResultPointerInt
}
/** Namespace for the script-runner free functions of the old ABI
*(`run_lua_script`, `run_js_script`, `run_script`).
*/
class Scripting internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class ScriptingCleaner(val handle: Pointer, val lib: ScriptingLib) : Runnable {
        override fun run() {
            lib.Scripting_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, Scripting.ScriptingCleaner(handle, Scripting.lib));
    }

    companion object {
        internal val libClass: Class<ScriptingLib> = ScriptingLib::class.java
        internal val lib: ScriptingLib = Native.load("nucleation", libClass)
        @JvmStatic
        
        /** Run a Lua script file. Returns the schematic the script assigns to
        *`result`; `NotFound` if it produced none, `Parse` if it failed, and
        *`InvalidArgument` when built without the `scripting-lua` feature.
        */
        fun runLuaScript(path: String): Result<Schematic> {
            val pathSliceMemory = PrimitiveArrayTools.borrowUtf8(path)
            
            val returnVal = lib.Scripting_run_lua_script(pathSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    val selfEdges: List<Any> = listOf()
                    val handle = nativeOkVal 
                    val returnOpaque = Schematic(handle, selfEdges, true)
                    return returnOpaque.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                pathSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** Run a JS script file. Returns the schematic the script assigns to
        *`result`; `NotFound` if it produced none, `Parse` if it failed, and
        *`InvalidArgument` when built without the `scripting-js` feature.
        */
        fun runJsScript(path: String): Result<Schematic> {
            val pathSliceMemory = PrimitiveArrayTools.borrowUtf8(path)
            
            val returnVal = lib.Scripting_run_js_script(pathSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    val selfEdges: List<Any> = listOf()
                    val handle = nativeOkVal 
                    val returnOpaque = Schematic(handle, selfEdges, true)
                    return returnOpaque.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                pathSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** Run a script file, auto-detecting the engine by extension (`.lua` or
        *`.js`). Returns the schematic the script assigns to `result`; `NotFound`
        *if it produced none, `Parse` if it failed (including unsupported
        *extensions).
        */
        fun runScript(path: String): Result<Schematic> {
            val pathSliceMemory = PrimitiveArrayTools.borrowUtf8(path)
            
            val returnVal = lib.Scripting_run_script(pathSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    val selfEdges: List<Any> = listOf()
                    val handle = nativeOkVal 
                    val returnOpaque = Schematic(handle, selfEdges, true)
                    return returnOpaque.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                pathSliceMemory.close()
            }
        }
    }

}

File: src/bridge/scripting.rs
//! Embedded scripting (Lua / JS) entry points. Port of `ffi/scripting.rs`.
//!
//! The module as a whole is gated in `bridge/mod.rs` behind
//! `any(feature = "scripting-lua", feature = "scripting-js")`, mirroring the old
//! per-function gating; within the module, the engine-specific methods keep a stable
//! signature and fail with `InvalidArgument` when their engine is compiled out
//! (feature-dependent behavior lives in the body, per the porting rules).

#[diplomat::bridge]
pub mod ffi {
    use super::super::schematic::ffi::Schematic;
    use super::super::shared::ffi::NucleationError;

    /// Namespace for the script-runner free functions of the old ABI
    /// (`run_lua_script`, `run_js_script`, `run_script`).
    #[diplomat::opaque]
    pub struct Scripting;

    impl Scripting {
        /// Run a Lua script file. Returns the schematic the script assigns to
        /// `result`; `NotFound` if it produced none, `Parse` if it failed, and
        /// `InvalidArgument` when built without the `scripting-lua` feature.
        pub fn run_lua_script(path: &DiplomatStr) -> Result<Box<Schematic>, NucleationError> {
            let path = std::str::from_utf8(path).map_err(|_| NucleationError::InvalidArgument)?;
            #[cfg(feature = "scripting-lua")]
            {
                match crate::scripting::lua_engine::run_lua_script(path) {
                    Ok(Some(ss)) => Ok(Box::new(Schematic(ss.inner))),
                    Ok(None) => Err(NucleationError::NotFound),
                    Err(_) => Err(NucleationError::Parse),
                }
            }
            #[cfg(not(feature = "scripting-lua"))]
            {
                let _ = path;
                Err(NucleationError::InvalidArgument)
            }
        }

        /// Run a JS script file. Returns the schematic the script assigns to
        /// `result`; `NotFound` if it produced none, `Parse` if it failed, and
        /// `InvalidArgument` when built without the `scripting-js` feature.
        pub fn run_js_script(path: &DiplomatStr) -> Result<Box<Schematic>, NucleationError> {
            let path = std::str::from_utf8(path).map_err(|_| NucleationError::InvalidArgument)?;
            #[cfg(feature = "scripting-js")]
            {
                match crate::scripting::js_engine::run_js_script(path) {
                    Ok(Some(ss)) => Ok(Box::new(Schematic(ss.inner))),
                    Ok(None) => Err(NucleationError::NotFound),
                    Err(_) => Err(NucleationError::Parse),
                }
            }
            #[cfg(not(feature = "scripting-js"))]
            {
                let _ = path;
                Err(NucleationError::InvalidArgument)
            }
        }

        /// Run a script file, auto-detecting the engine by extension (`.lua` or
        /// `.js`). Returns the schematic the script assigns to `result`; `NotFound`
        /// if it produced none, `Parse` if it failed (including unsupported
        /// extensions).
        pub fn run_script(path: &DiplomatStr) -> Result<Box<Schematic>, NucleationError> {
            let path = std::str::from_utf8(path).map_err(|_| NucleationError::InvalidArgument)?;
            match crate::scripting::run_script(path) {
                Ok(Some(ss)) => Ok(Box::new(Schematic(ss.inner))),
                Ok(None) => Err(NucleationError::NotFound),
                Err(_) => Err(NucleationError::Parse),
            }
        }
    }
}
